File: wizard-py/test_parser.py
# Copyright 2020 Google LLC. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


import ast
import constants


def get_test_methods(test_path):
    with open(test_path, 'r') as f:
        content = "".join(f.readlines())
        parsed_nodes = ast.iter_child_nodes(ast.parse(content))
        parsed_nodes = [x for x in parsed_nodes]

        # Handle test methods wrapped in classes
        class_nodes = [x for x in parsed_nodes if 'ClassDef' in str(type(x))]
        test_nodes = [x for x in parsed_nodes if x not in class_nodes]

        for c in class_nodes:
            test_nodes += c.body

        test_nodes = [x for x in test_nodes if hasattr(x, 'name')]
        test_nodes = [x for x in test_nodes if x.name.startswith('test_')]

        for n in test_nodes:
            n.test_path = test_path

        # Verify file contains no duplicate method names
        # (Only relevant for test methods wrapped in classes)
        used_test_names = set()
        for n in test_nodes:
            if n.name in used_test_names:
                raise ValueError(
                    f'Test name {n.name} in file {test_path} must be unique.')
            used_test_names.add(n.name)

        return test_nodes


def get_test_to_method_map(test_methods):
    test_method_map = {}

    def __recursor__(expr):
        type_str = str(type(expr))
        is_func = hasattr(expr, 'func')

        if '.Attribute' in type_str and hasattr(expr.value, 'id'):
            # Direct method invocation
            return [{
                'class_name': expr.value.id,
                'method_name': expr.attr
            }]

        if is_func and hasattr(expr.func, 'value'):
            # HTTP-route invoked methods
            # (both flask and webapp2)
            func = expr.func

            if hasattr(func.value, 'id') and \
               func.value.id in constants.HTTP_CLASS_NAMES and \
               func.attr in constants.HTTP_METHOD_NAMES and \
               hasattr(expr.args[0], 's'):
                return [{
                    'url': expr.args[0].s,
                    'http_method': func.attr
                }]

        if hasattr(expr, 'value'):
            return __recursor__(expr.value)

        if is_func:
            return __recursor__(expr.func)

        if hasattr(expr, 'test') and hasattr(expr.test, 'comparators'):
            results = []
            results += __recursor__(expr.test.comparators[0])
            results += __recursor__(expr.test.left)
            return results

        if '.With' in type_str or '.For' in type_str:
            results = []
            for x in expr.body:
                results += [x for x in __recursor__(x) if x]

            return results  # may contain duplicates

        return []

    for method in test_methods:
        for x in method.body:
            rList = [r for r in __recursor__(x) if r]
            for r in rList:
                rKey = (
                    r.get('http_method', r.get('class_name')),
                    r.get('url', r.get('method_name'))
                )
                if rKey not in test_method_map:
                    test_method_map[rKey] = []
                test_method_map[rKey].append(
                    (method.test_path, method.name))

    return test_method_map


def add_method_names_to_tests(source_methods, test_method_map):
    for m in source_methods:
        d = m.drift
        if 'test_methods' not in m.drift:
            m.drift['test_methods'] = []

        keys = []
        if d['parser'] == 'direct_invocation':
            keys = [(d.get('class_name'), d.get('method_name'))]
        elif d['parser'] == 'webapp2_router':
            keys = [(d.get('http_method'), d.get('url'))]
        elif d['parser'] == 'flask_router':
            keys = [(x, d.get('url')) for x in d['http_methods']]

        for key in keys:
            if key in test_method_map:
                if test_method_map[key]:
                    m.drift['test_methods'] += test_method_map[key]

File: wizard-js/test-parser.js
// Copyright 2020 Google LLC. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

'use strict';

const {
    EXPRESS_METHOD_NAMES,
    REGION_TAG_CONSTANT_REGEX
} = require('./constants')

const uniq = require('uniq');
const deepEqual = require('deep-equal');
const path = require('path');

const parsers = {
    cloudFunction: require('./test-parsers/core/cloud-functions.js'),
    httpRequest: require('./test-parsers/core/http-requests.js'),
    directInvocation: require('./test-parsers/core/direct-invocation.js'),
    execSync: require('./test-parsers/core/exec-sync.js'),
    regionTagInConstant: require('./test-parsers/extra/region-tag-in-constant.js')
};

exports.getDescribeClauses = (testTokens) => {
    const clauses = testTokens.body.filter(s => s.expression && s.expression.callee && s.expression.callee.name === 'describe');
    return clauses;
};

const parseItClause = (itClause, describeChain, testPath, topLevelRegionTag) => {
    // Ignore non-"it" clauses (e.g. "before", "after")
    if (!itClause.expression || !itClause.expression.callee || itClause.expression.callee.name !== 'it') {
        return {};
    }

    // Process "it" clauses
    // ASSUMPTION: test is always 2nd argument of "it" clause
    const itClauseArgs = itClause.expression.arguments;

    // Compute final describe chain
    describeChain = describeChain.concat(itClauseArgs[0].value);

    let testStatements = itClauseArgs[1].body.body;
    const testPaths = [testPath];
    const commandToDescribeMap = {}

    // Collapse "try" statements
    testStatements = testStatements.map(t => t.block && t.block.body[0] || t)

    // Parse tests
    parsers.httpRequest.parse(testStatements, commandToDescribeMap, describeChain, testPaths);
    parsers.cloudFunction.parse(testStatements, commandToDescribeMap, describeChain, testPaths);
    parsers.directInvocation.parse(testStatements, commandToDescribeMap, describeChain, testPaths);
    parsers.execSync.parse(testStatements, commandToDescribeMap, describeChain, testPaths);
    parsers.regionTagInConstant.parse(
        testStatements, 
        commandToDescribeMap, 
        describeChain, 
        testPaths,
        topLevelRegionTag);

    return commandToDescribeMap;
}

exports.getDescribeChainMap = (describeClauses, testPath, topLevelRegionTag, describeChain) => {
    describeChain = describeChain || [];

    // Parse child describe clauses
    const aggregateCliMap = {};
    describeClauses.forEach(c => {
        let resultCliMap;
        if (c.expression && c.expression.callee && c.expression.callee.name === 'describe') {
            // ASSUMPTION:
            //   description will always be 1st argument of describe()
            //   test function will always be 2nd argument of describe()
            const children = c.expression.arguments[1].body.body;

            let description = c.expression.arguments[0];

            // Handle "region tag in constant" tests (e.g. nodejs-translate)
            description = description.value || description.name;
            if (topLevelRegionTag && description.match(REGION_TAG_CONSTANT_REGEX)) {
                description = topLevelRegionTag;
            }

            const newDescribeChain = describeChain.concat(description);
            resultCliMap = exports.getDescribeChainMap(
                children, 
                testPath, 
                topLevelRegionTag, 
                newDescribeChain);
        } else {
            resultCliMap = parseItClause(c, describeChain, testPath, topLevelRegionTag);
        }

        // Aggregate recursive results
        Object.keys(resultCliMap).forEach(key => {
            aggregateCliMap[key] = (aggregateCliMap[key] || []).concat(resultCliMap[key]);
        });
    }); 

    return aggregateCliMap;
}

exports.addDescribeChainToMethods = (testPath, methodsWithDriftData, describeChainMap, fileWideInvocation) => {
    // Initialize default values used in test parsing
    methodsWithDriftData.map(m => m.drift).forEach(d => {
        d.testData = d.testData || [];
    })

    methodsWithDriftData.forEach(method => {
        const d = method.drift;

        let describeObjs;
        const sourcePathNoExtension = path.parse(d.sourcePath).name.toLowerCase();
        if (d.cliInvocation) {
            describeObjs = describeChainMap[d.cliInvocation.toLowerCase()]
        } else if (d.urlPath && describeChainMap[d.urlPath.toLowerCase()]) {
            describeObjs = describeChainMap[d.urlPath.toLowerCase()][d.httpMethod.toLowerCase()];
        } else if (d.name && describeChainMap[d.name.toLowerCase()]) {
            // Directly-invoked method (or Cloud Function)
            describeObjs = describeChainMap[d.name.toLowerCase()]
        } else if (d.functionName && describeChainMap['/' + d.functionName.toLowerCase()]) {
            // URL-invoked Cloud Function
            // (Ignores HTTP methods!)
            const mapEntry = describeChainMap['/' + d.functionName.toLowerCase()];
            describeObjs = {};
            Object.values(mapEntry).map(v => Object.assign(describeObjs, v))
        } else if (path.basename(d.sourcePath) === fileWideInvocation) {
            // Single-test test files
            describeObjs = Object.values(describeChainMap)[0]
        } else if (describeChainMap[sourcePathNoExtension]) {
            // Single-snippet source files, referenced by name (e.g. `snippetName.js`)
            describeObjs = describeChainMap[sourcePathNoExtension]
        } else if (d.sourcePath.includes('quickstart')) {
            describeObjs = describeChainMap[d.sourcePath.toLowerCase()];
        }

        if (describeObjs) {
            d.testData = d.testData.concat(describeObjs);
        }
    })

    // Remove duplicates
    methodsWithDriftData.map(m => m.drift).forEach(d => {
        d.testData = uniq(d.testData, (x, y) => deepEqual(x, y) ? 0 : 1);
    })
}
